### PR TITLE
Python to C Difference Equation Generation

### LINKED ISSUE #1449: Python Difference Equation to C Generation
### Description of the task

For our low-level wheel controllers we need to translate the difference equation from the python implementation (where the controller design is tested and validated) to our C firmware implementation so it can run on the robot.

We should generate a header file, controller.h that contains this generated code.

The difference equation is made up of coefficients (determined by python code) and static multiplication of present and past data.

An example would look as follows:

y_0 = A*x_0 + Bx_1 + C*x_2 + D_y_1 + E_y_2 ....

Where capital letters are static coefficients and x_# is the #'s oldest item in a circular buffer of previous inputs and y_# is the #'s oldest item in a circular buffer of previous outputs.
### Acceptance criteria

- The C code is generated by a bazel rule to controller.h
- The generation can handle any number of coefficients and circular buffer length
- Circular buffers initialize themselves to all zeros on startup

### PR DESCRIPTION
**Pull request by kelvinkoon**
### Description

Generating C code using a Python file for the Difference Equation.

### Testing Done

Test for `generate_difference_equation.py` found in `generate_difference_equation_test.py`. 
Also manually tested using genrule and checked corresponding result.

### Resolved Issues

Resolves #1449 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by kelvinkoon** [comment]
Outstanding questions about `controller.h`:
- What is the expected plan to take in the static coefficients (i.e A, B, C, etc.)?
- When referring to "x_#" and "y_#", does that mean we compute using all the elements in the buffer with 0 being the newest? If so, I'm thinking of adding an additional method to `circular_buffer.h` that will return an array of the values from head to tail to be used for this. Please let me know if there's better implementations/ideas of going about this.

**Review by EvanMorcom — COMMENTED** [review]
Looks like you've got all the functions we'll need 🔥 🧨

**Comment by jonathanlew** [comment]
> Outstanding questions about `controller.h`:
> 
>     * What is the expected plan to take in the static coefficients (i.e A, B, C, etc.)?
> 
>     * When referring to "x_#" and "y_#", does that mean we compute using all the elements in the buffer with 0 being the newest? If so, I'm thinking of adding an additional method to `circular_buffer.h` that will return an array of the values from head to tail to be used for this. Please let me know if there's better implementations/ideas of going about this.

I think it would be similar to dynamic params in that we have the coefficients in a .yaml file that gets generated into a header file that gets included in the controller. @EvanMorcom thoughts?

**Review by garethellis0 — COMMENTED** [review]
Hmm, so I think we should aim for having the smallest amount of code generated, just because it makes it easier to maintain. Here's something like what I'm thinking:
```
#define WHEEL_SPEED_CONTROLLER_TICK_TIME 1e7
#define GENERATED_NUM_WHEEL_SPEED_COEFFICIENTS;
#define GENERATED_NUM_PREVIOUS_WHEEL_SPEED_COMMAND_COEFFICIENTS;

float GENERATED_WHEEL_SPEED_COEFFICIENTS[GENERATED_NUM_WHEEL_SPEED_COEFFICIENTS] = 
        1, 2, 3
    };
float GENERATED_PREVIOUS_WHEEL_SPEED_COMMAND_COEFFICIENTS[
    GENERATED_NUM_PREVIOUS_WHEEL_SPEED_COMMAND_COEFFICIENTS] = 
        4, 5, 6
    };
```
I don't think it should need to be too much more then this?

**Comment by kelvinkoon** [comment]
> Hmm, so I think we should aim for having the smallest amount of code generated, just because it makes it easier to maintain. Here's something like what I'm thinking:
> 
> ```
> #define WHEEL_SPEED_CONTROLLER_TICK_TIME 1e7
> #define GENERATED_NUM_WHEEL_SPEED_COEFFICIENTS;
> #define GENERATED_NUM_PREVIOUS_WHEEL_SPEED_COMMAND_COEFFICIENTS;
> 
> float GENERATED_WHEEL_SPEED_COEFFICIENTS[GENERATED_NUM_WHEEL_SPEED_COEFFICIENTS] = 
>         1, 2, 3
>     };
> float GENERATED_PREVIOUS_WHEEL_SPEED_COMMAND_COEFFICIENTS[
>     GENERATED_NUM_PREVIOUS_WHEEL_SPEED_COMMAND_COEFFICIENTS] = 
>         4, 5, 6
>     };
> ```
> 
> I don't think it should need to be too much more then this?

The intent of it was to I think check that this is the code we want to generate before moving on to writing the Python script to attempt to generate the code just to make sure I know what I'm generating is right. 
I think in the final non-WIP version, `controller.h` will hopefully look something like that.

**Comment by garethellis0** [comment]
> we have the coefficients in a .yaml file that gets generated into a header file that gets included in the controller

I think it would more be that the python version of the controller would be able to generate a header file containing the coefficients (ie. generating effectively what I have above).

**Comment by garethellis0** [comment]
> I think in the final non-WIP version, controller.h will hopefully look something like that.

@kelvinkoon - sorry, what do you mean by "that" here? Just to be totally clear, what I put above is what I think we probably want to be generating.

**Comment by kelvinkoon** [comment]
> > I think in the final non-WIP version, controller.h will hopefully look something like that.
> 
> @kelvinkoon - sorry, what do you mean by "that" here? Just to be totally clear, what I put above is what I think we probably want to be generating.

Oh, I see. So with generating, I don't need those functions? My understanding was that I'd be generating functions to do things like add values to the buffers, clear buffers, etc. and this was just for planning out that.

**Comment by garethellis0** [comment]
Yeah, I don't think any of the buffer stuff needs to be generated. We really only need to generate the things that would change if we were to update the python controller, namely the coefficients. The buffers shouldn't really need to change if we regenerate code, and generated code is hard to maintain (hence we should really try to minimise it).

**Comment by kelvinkoon** [comment]
> Yeah, I don't think any of the buffer stuff needs to be generated. We really only need to generate the things that would change if we were to update the python controller, namely the coefficients. The buffers shouldn't really need to change if we regenerate code, and generated code is hard to maintain (hence we should really try to minimise it).

I've added a file called `generate_difference_equation.py` which should generate `CONTROLLER_DIFFERENCE_EQUATION_COEFFICIENTS`. File is found inside `firmware/app/control`, but when I try building it, I get an error saying `No module named 'pyparsing'`. 
If someone could check over if the file and location look right, would be appreciated. After that, I can try to hook it to Bazel's genrule to generate the code and then write tests for it.

**Comment by garethellis0** [comment]
Perhaps add `pyparsing` as another requirement?

Summoning the :snake: tamer for his opinion: @MathewMacDougall

**Comment by MathewMacDougall** [comment]
Looking at this now and I can't reproduce the failure locally. building `generate_difference_equation` works fine for me, but it looks like I have `pyparsing` in my global python for some reason.

Yes adding `pyparsing` as a requirement will probably fix this, although I can't really see where it's used (likely a transitive dependency).

To add `pyparsing`
* add `pyparsing==2.2.0` to the python control deps `requirements.txt` file
* add `requirement("pyparsing")` to the deps of the target

**Comment by kelvinkoon** [comment]
> Looking at this now and I can't reproduce the failure locally. building `generate_difference_equation` works fine for me, but it looks like I have `pyparsing` in my global python for some reason.
> 
> Yes adding `pyparsing` as a requirement will probably fix this, although I can't really see where it's used (likely a transitive dependency).
> 
> To add `pyparsing`
> 
>     * add `pyparsing==2.2.0` to the python control deps `requirements.txt` file
> 
>     * add `requirement("pyparsing")` to the deps of the target

Adding `pyparsing`, `cycler`, and `kiwisolver` to the BUILD file did the trick. Thanks guys!
Would now be the right time to hook it up to Bazel genrule to try it? 
Also, the `generate_difference_equation` file is currently writing to itself. Is there a preferred directory to write it to?

File: src/firmware/app/control/generate_difference_equation.py
Comment (MathewMacDougall) [inline src/firmware/app/control/generate_difference_equation.py]: nit: missing newline at EOF

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Going to defer to @EvanMorcom and @garethellis0 as to the best place for the generated file.

**Comment by garethellis0** [comment]
> Going to defer to @EvanMorcom and @garethellis0 as to the best place for the generated file.

@kelvinkoon - IMO this should all go in `src/firmware/app/control/python_controller`

File: src/firmware/app/control/generate_difference_equation.py
Comment (kelvinkoon) [reply src/firmware/app/control/generate_difference_equation.py]: Fixed.

**Comment by kelvinkoon** [comment]
Generated code now looks like what @garethellis0 posted earlier (hard-coded values are passed into the function ATM).
```
#define WHEEL_SPEED_CONTROLLER_TICK_TIME 1e7
#define GENERATED_NUM_WHEEL_SPEED_COEFFICIENTS 4;
#define GENERATED_NUM_PREVIOUS_WHEEL_SPEED_COMMAND_COEFFICIENTS 7;
float GENERATED_WHEEL_SPEED_COEFFICIENTS[GENERATED_NUM_WHEEL_SPEED_COEFFICIENTS] =
{1, 1};
float GENERATED_PREVIOUS_WHEEL_SPEED_COMMAND_COEFFICIENTS[GENERATED_NUM_PREVIOUS_WHEEL_SPEED_COMMAND_COEFFICIENTS] =
{1, 1, 1};
```

According to @EvanMorcom , having the `generate` function in `generate_difference_equation.py` is enough for the scope of the ticket since passing in the discrete function will be relatively straightforward when the controller design file is finished. So I suppose that would just leave passing CI and tests...?

**Comment by garethellis0** [comment]
Yep!

**Comment by kelvinkoon** [comment]
PR is no longer WIP. Added a unit test and code generation seems to work with the hard-coded values.

File: src/firmware/app/control/python_controller/BUILD
Comment (MathewMacDougall) [inline src/firmware/app/control/python_controller/BUILD]: this should probably be named more descriptively so consumer know it generates the controller coefficients. Maybe something like `controller_coefficients_cc_generate`

**Review by garethellis0 — DISMISSED** [review]

File: src/firmware/app/control/python_controller/generate_difference_equation.py
Comment (MathewMacDougall) [inline src/firmware/app/control/python_controller/generate_difference_equation.py]: Where does 1e7 come from? What are the units?

File: src/firmware/app/control/python_controller/generate_difference_equation.py
Comment (MathewMacDougall) [inline src/firmware/app/control/python_controller/generate_difference_equation.py:14]: Why do some variables have single `{}` and some have triple `{}`?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/firmware/app/control/python_controller/generate_difference_equation.py
Comment (kelvinkoon) [reply src/firmware/app/control/python_controller/generate_difference_equation.py]: `WHEEL_SPEED_CONTROLLER_TICK_TIME` is the period we expect the controller to run in. 1e7 was just taken from the sample snippet @garethellis0 posted above. Should be in seconds, will add a comment clarifying that.

File: src/firmware/app/control/python_controller/generate_difference_equation.py
Comment (kelvinkoon) [reply src/firmware/app/control/python_controller/generate_difference_equation.py:14]: `numerator_length` and `denominator_length` need only one bracket because we're just replacing them with the `format` method.
`numerator_coefficients` and `denominator_coefficients` need 3 brackets. This is because we need one pair of brackets for the `format` method. Another pair of brackets is because we need to initialize the array (`int arr[3] = {1,2,3}` for instance). The last pair is like an 'escape' for the pair of brackets. 
tl;dr one pair to do the format, one pair to initialize array, one pair to escape

File: src/firmware/app/control/python_controller/generate_difference_equation.py
Comment (MathewMacDougall) [reply src/firmware/app/control/python_controller/generate_difference_equation.py]: So this is an example for now, or a reasonably "real" value? Either way, seems maybe like something the python controller or something should define rather than being "hidden" in the generated code template? @garethellis0 @EvanMorcom 

Also I'm assuming that should be `1e-7` then?

Comment (kelvinkoon) [reply src/firmware/app/control/python_controller/generate_difference_equation.py]: I could have `generate` take in a parameter and generate it instead. Let me know if that's preferred.

File: src/firmware/app/control/python_controller/BUILD
Comment (kelvinkoon) [reply src/firmware/app/control/python_controller/BUILD]: Done.

File: src/firmware/app/control/python_controller/generate_difference_equation.py
Comment (jonathanlew) [reply src/firmware/app/control/python_controller/generate_difference_equation.py:14]: can we add a comment to that effect?

**Review by jonathanlew — COMMENTED** [review]
looking good.

File: src/firmware/app/control/python_controller/generate_difference_equation.py
Comment (kelvinkoon) [reply src/firmware/app/control/python_controller/generate_difference_equation.py:14]: Done.

**Review by jonathanlew — DISMISSED** [review]
LGTM pending Mat's comment

File: src/firmware/app/control/python_controller/generate_difference_equation.py
Comment (MathewMacDougall) [reply src/firmware/app/control/python_controller/generate_difference_equation.py]: @garethellis0 @EvanMorcom bumping this

Comment (garethellis0) [reply src/firmware/app/control/python_controller/generate_difference_equation.py]: @EvanMorcom - is there a function that @kelvinkoon can use from the model to get the tick time it was based on?

Comment (akhilveeraghanta) [reply src/firmware/app/control/python_controller/generate_difference_equation.py]: @EvanMorcom bumping this

Comment (EvanMorcom) [reply src/firmware/app/control/python_controller/generate_difference_equation.py]: Ah yes. There is a member if the discrete TF class in the python control toolbox called 'dt' that gives this period. 
Screen Shot 2020-07-02 at 10 03 32 AM

**Review by EvanMorcom — CHANGES_REQUESTED** [review]
LGTM :fire: 🚒  🧨 

Only request is incorporating the interpolation period from the discrete TD

File: src/firmware/app/control/python_controller/generate_difference_equation.py
Comment (kelvinkoon) [reply src/firmware/app/control/python_controller/generate_difference_equation.py]: Added.

**Review by MathewMacDougall — APPROVED** [review]
LGTM pending Evan's comment

**Comment by kelvinkoon** [comment]
Pinging @jonathanlew and @EvanMorcom for reviews with respective comments addressed.

**Review by jonathanlew — APPROVED** [review]
approved pending Evan's review

**Review by EvanMorcom — APPROVED** [review]
LGTM :fire:

**Pull request merged**
